managed_transform_buffer/src/managed_transform_buffer_provider.cpp

```diff
@@ -0,0 +1,385 @@
+  tf2::Duration cache_time)
+{
+  static ManagedTransformBufferProvider instance(clock_type, discovery_timeout, cache_time);
+  if (force_dynamic) {
+    if (instance.isStatic()) {
+      std::lock_guard<std::mutex> listener_lock(instance.listener_mutex_);
+      std::unique_lock<std::shared_mutex> unq_buffer_lock(instance.buffer_mutex_);
+      std::unique_lock<std::shared_mutex> unq_tree_lock(instance.tree_mutex_);
+      instance.registerAsDynamic();
+    }
+  }
+  if (clock_type != instance.clock_->get_clock_type()) {
+    RCLCPP_WARN_THROTTLE(
+      instance.logger_, *instance.clock_, 3000,
+      "Input clock type does not match (%d vs. %d). Input clock type will be ignored.", clock_type,
+      instance.clock_->get_clock_type());
+  }
+  return instance;
+}
+
+std::optional<TransformStamped> ManagedTransformBufferProvider::getTransform(
+  const std::string & target_frame, const std::string & source_frame, const tf2::TimePoint & time,
+  const tf2::Duration & timeout, const rclcpp::Logger & logger)
+{
+  return get_transform_(target_frame, source_frame, time, timeout, logger);
+}
+
+bool ManagedTransformBufferProvider::isStatic() const
+{
+  return is_static_.load();
+}
+
+ManagedTransformBufferProvider::ManagedTransformBufferProvider(
+  rcl_clock_type_t clock_type, tf2::Duration discovery_timeout, tf2::Duration cache_time)
+: clock_(std::make_shared<rclcpp::Clock>(clock_type)),
+  discovery_timeout_(discovery_timeout),
+  logger_(rclcpp::get_logger("ManagedTransformBuffer"))
+{
+  executor_ = std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
+  executor_thread_ = std::make_shared<std::thread>(
+    std::bind(&rclcpp::executors::SingleThreadedExecutor::spin, executor_));
+
+  static_tf_buffer_ = std::make_unique<TFMap>();
+  tf_tree_ = std::make_unique<TreeMap>();
+  tf_buffer_ = std::make_unique<tf2_ros::Buffer>(clock_, cache_time);
+  tf_buffer_->setUsingDedicatedThread(true);
+
+  tf_options_ = tf2_ros::detail::get_default_transform_listener_sub_options();
+  tf_static_options_ = tf2_ros::detail::get_default_transform_listener_static_sub_options();
+  cb_ = std::bind(&ManagedTransformBufferProvider::tfCallback, this, std::placeholders::_1, false);
+  cb_static_ =
+    std::bind(&ManagedTransformBufferProvider::tfCallback, this, std::placeholders::_1, true);
+  options_.start_parameter_event_publisher(false);
+  options_.start_parameter_services(false);
+  random_engine_ = std::mt19937(std::random_device{}());
+  dis_ = std::uniform_int_distribution<>(0, 0xFFFFFF);
+  registerAsUnknown();
+}
+
+ManagedTransformBufferProvider::~ManagedTransformBufferProvider()
+{
+  executor_->cancel();
+  if (executor_thread_->joinable()) {
+    executor_thread_->join();
+  }
+}
+
+void ManagedTransformBufferProvider::activateListener()
+{
+  std::lock_guard<std::mutex> listener_lock(listener_mutex_);
+  operational_threads_.fetch_add(1);
+  if (operational_threads_.load() != 1) return;
```

Comment:
nit: does `operational_threads_` need to be an integer? Is the logic not more clear using a bool?